src/twisted/protocols/_smb/_base.py

```diff
@@ -0,0 +1,396 @@
+    """a universial unique ID"""
+    default = _conv_uuid(default)
+    return attr.ib(
+        default=default,
+        metadata={SMB_METADATA: "16s"},
+        type=uuid_mod.UUID,
+        converter=_conv_uuid,
+        validator=default_only if locked else None,
+    )
+
+
+def _conv_uuid(x: Union[str, bytes, uuid_mod.UUID]) -> uuid_mod.UUID:
+    if type(x) is str:
+        return uuid_mod.UUID(x)
+    elif type(x) is bytes:
+        return uuid_mod.UUID(bytes_le=x)
+    else:
+        return cast(uuid_mod.UUID, x)
+
+
+def pack(obj: Any) -> bytes:
+    """
+    pack an object into binary data. The object must have been decorated
+    with L{attr.s} and the fields set with the appropriate metadata using
+    the helpers in this module
+
+    @rtype: L{bytes}
+    """
+    strct = _get_struct(type(obj))
+    args = tuple(_conv_arg(obj, i) for i in smb_fields(type(obj)))
+    log.debug("about to pack {strct!r} {args}", strct=strct, args=args)
+    return strct.pack(*args)
+
+
+def _conv_arg(obj: Any, attrib: attr.Attribute) -> Any:
+    val = getattr(obj, attrib.name)
+    if type(val) is uuid_mod.UUID:
+        val = val.bytes_le
+    if isinstance(val, enum.Enum):
+        val = val.value
+    return val
+
+
+IGNORE = 0
+ERROR = 1
+OFFSET = 2
+DATA = 3
+
+
+def unpack(cls: Any, data: bytes, offset: int = 0, remainder: int = IGNORE) -> Any:
+    """
+    unpack binary data into an object.
+
+    @param cls: the class, must be decorated with L{attr.s} and have
+    members with the appropriate metadata using the helpers from this
+    module.
+    @type cls: L{type}
+
+    @param data: the data to unpack
+    @type data: L{bytes}
+
+    @param remainder: what to do with remaining data if longer than required
+    to fill C{cls}
+     - C{IGNORE} ignore it
+     - C{ERROR} throw a L{SMBError}
+     - C{OFFSET} return offset into data
+     where remainder begins as second item of tuple
+     - C{DATA} return remaining data as second item of tuple
+    @type remainder: L{int}
+
+    @param offset: offset into data to begin from
+    @type offset: L{int}
+
+    @return: an instance of C{cls}, or a 2-tuple, first item the former,
+             second as determined by C{remainder}
+    """
+    strct = _get_struct(cls)
+    if remainder == ERROR and strct.size + offset < len(data):
+        raise SMBError("unexpected remaining data")
+    ret = strct.unpack_from(data, offset=offset)
+    fields = smb_fields(cls)
+    assert len(fields) == len(ret)
+    kwargs = {}
+    for i in range(len(ret)):
+        val = ret[i]
+        if fields[i].type is uuid_mod.UUID:
+            val = uuid_mod.UUID(bytes_le=val)
+        kwargs[fields[i].name] = val
+    obj = cls(**kwargs)
+    if remainder <= ERROR:
+        return obj
+    elif remainder == OFFSET:
+        return (obj, offset + strct.size)
+    else:
+        return (obj, data[offset + strct.size :])
+
+
+def smb_fields(cls: Any) -> List[Any]:
```

Comment:
coding standard: we don't use underscores like PEP8 in Twisted
```suggestion
def smbFields(cls: Type[object]) -> List[Any]:
```